bip-PSBT-SP.mediawiki

```diff
@@ -0,0 +1,196 @@
+<pre>
+  BIP: ?
+  Layer: Applications
+  Title: Sending Silent Payments with PSBTs
+  Author: Andrew Toth <andrewstoth@gmail.com>
+          Ava Chow <me@achow101.com>
+          josibake <josibake@protonmail.com>
+  Comments-Summary: No comments yet.
+  Comments-URI: TBD
+  Status: Draft
+  Type: Standards Track
+  Created: 2024-05-14
+  License: BSD-2-Clause
+  Post-History: https://groups.google.com/g/bitcoindev/c/5G5wzqUXyk4
+  Requires: 352, 370
+</pre>
+
+==Introduction==
+
+===Abstract===
+
+This document proposes additional fields and updated role responsibilities for BIP370 PSBTv2
+which adds support for sending to silent payments as described in BIP352.
+
+===Copyright===
+
+This BIP is licensed under the 2-clause BSD license.
+
+===Motivation===
+
+Partially Signed Bitcoin Transaction Version 2 as described in BIP370 is not compatible with sending to silent payments as described in BIP352. 
+In particular, the output script of a silent payment cannot be computed until after all transaction inputs have been added.
+Additionally, the silent payment outputs computed by a signer must be verifiable by other entities, otherwise funds could be sent to an incorrect output script.
+Therefore, new fields and role responsibilities must be added to carry, compute, and verify the silent payment data.
+
+==Specification==
+
+This document specifies new fields and new field inclusion/exclusion requirements.
+
+The new global types are defined as follows:
+
+{|
+! Name
+! <tt><keytype></tt>
+! <tt><keydata></tt>
+! <tt><keydata></tt> Description
+! <tt><valuedata></tt>
+! <tt><valuedata></tt> Description
+! Versions Requiring Inclusion
+! Versions Requiring Exclusion
+! Versions Allowing Inclusion
+|-
+| Silent Payment Global ECDH Share
+| <tt>PSBT_GLOBAL_SP_ECDH_SHARE = 0x07</tt>
+| <tt><33 byte scan key> <36 byte outpoint>*</tt>
+| The scan key and a list of outpoints corresponding to the prevouts of the inputs that this ECDH share is for. The outpoints are composed of a 32 byte txid followed by a 32-bit little endian uint.
+| <tt><33 byte share></tt>
+| An ECDH share for a scan key, followed by a list of outpoints. The ECDH shared is computed with ''a * B_scan'', where ''a'' is the sum of all private keys of the inputs matching the list of outpoints, and ''B_scan'' is the scan key of a recipient.
```

Comment:
I have an implementation question about this. If all the shares are `a * B_scan`, then how can the last signer (or the transaction extractor) calculate the output key? BIP-0352 mentions `input_hash = hash_BIP0352/Inputs(outpointL || A)`, but how would I extract `A` from the shares? Is it as simple as subtracting `B_scan` from the share?

Reply:
Ah, I guess a PSBT is supposed to contain all BIP-32 derivation information for all inputs. So `A` can be derived from the sum of all public keys. Is that correct? Then perhaps that could be mentioned here.

Reply:
> how would I extract A from the shares?

`A` in the context of the `input_hash` is `a*G`, or the sum of all public keys. This might be easier to understand from the receiver side https://github.com/bitcoin/bips/blob/master/bip-0352.mediawiki#scanning.

Reply:
> a PSBT is supposed to contain all BIP-32 derivation information for all inputs

I'm not sure I follow why we need BIP-32 derivation information? The public keys can be extracted by the prevout scripts of the inputs. https://github.com/bitcoin/bips/blob/master/bip-0352.mediawiki#inputs-for-shared-secret-derivation

Reply:
I think I understand, for output generation:

`Let ecdh_shared_secret = input_hash·a·Bscan`

replace `a·Bscan` with the ECDH share.

This is detailed [here](https://github.com/bitcoin/bips/pull/1687/files#diff-b41d8882b5f3bcf8086d19f64655573bdb4666f5091054c4b9509b60e185293aR174).

Reply:
True, the `PSBT_IN_HASH160` should probably work as well. I just haven't seen this field being set "in the wild" up to now, so not sure if all wallets will set it correctly (at least `lnd` doesn't, which perhaps means it's not fully BIP-0174 compliant?).
So I thought the public key in the BIP-32 derivation fields is probably more likely to be set in most wallets.
But perhaps this BIP doesn't have to go into details around that and can just mention something along the lines of:
"Signers creating the output keys will not be able to extract public keys from the witness as at that point the inputs aren't signed yet. The meta information present in the PSBT should include the public key of each input though, so creating the sum of all input public keys (`A`) should happen based on that information".

Reply:
> But perhaps this BIP doesn't have to go into details around that

Hmm I'm not sure that's a good idea to leave that unspecified. Different implementations could then depend on different metadata fields that may or may not be set and be incompatible. The purpose of a standard is to avoid that problem.

If it's easier we can use the BIP-32 derivation field since I see `PSBT_IN_HASH160` is not implemented in https://github.com/bitcoinjs/bip174 either. The difference is that the value of that field should be set to a fingerprint and derivation path which some signers might not want to share. We can specify in this document to set this field for non-taproot sp eligible inputs, and a zero fingerprint and empty path array for the value.